contrib/mesos/pkg/proc/adapter.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package proc
+
+import (
+	"fmt"
+)
+
+type processAdapter struct {
+	Process
+	delegate Doer
+}
+
+func (p *processAdapter) Do(a Action) <-chan error {
+	errCh := NewErrorOnce(p.Done())
+	go func() {
+		errOuter := p.Process.Do(func() {
+			errInner := p.delegate.Do(a)
+			errCh.forward(errInner)
+		})
+		// if the outer err is !nil then either the parent Process failed to schedule the
+		// the action, or else it backgrounded the scheduling task.
+		if errOuter != nil {
+			errCh.forward(errOuter)
+		}
+	}()
+	return errCh.Err()
+}
+
+// DoWith returns a process that, within its execution context, delegates to the specified Doer.
+// Expect a panic if either the given Process or Doer are nil.
+func DoWith(other Process, d Doer) Process {
+	if other == nil {
+		panic(fmt.Sprintf("cannot DoWith a nil process"))
```

Comment:
passing in a nil to this should fail fast. it's a programming error and the behavior is doc'd above.